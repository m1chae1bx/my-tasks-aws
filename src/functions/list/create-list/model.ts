import { JSONSchemaType, ajv } from "/opt/nodejs/util";

export interface CreateListRequest {
  pathParameters: {
    userId: string;
  };
  body: {
    name: string;
    isDefault?: boolean;
  };
}

export const createListRequestSchema: JSONSchemaType<CreateListRequest> = {
  $schema: "http://json-schema.org/draft-07/schema#",
  type: "object",
  properties: {
    pathParameters: {
      type: "object",
      properties: {
        userId: {
          type: "string",
        },
      },
      required: ["userId"],
    },
    body: {
      type: "object",
      properties: {
        name: {
          type: "string",
          maxLength: 100,
        },
        isDefault: {
          type: "boolean",
          nullable: true,
        },
      },
      required: ["name"],
    },
  },
  required: ["pathParameters", "body"],
};

export const validateRequest = ajv.compile(createListRequestSchema);
