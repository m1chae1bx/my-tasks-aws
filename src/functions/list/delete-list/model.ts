import { JSONSchemaType, ajv } from "/opt/nodejs/util";

export interface DeleteListRequest {
  pathParameters: {
    userId: string;
    listId: string;
  };
}

export const deleteListRequestSchema: JSONSchemaType<DeleteListRequest> = {
  $schema: "http://json-schema.org/draft-07/schema#",
  type: "object",
  properties: {
    pathParameters: {
      type: "object",
      properties: {
        userId: {
          type: "string",
        },
        listId: {
          type: "string",
        },
      },
      required: ["userId", "listId"],
    },
  },
  required: ["pathParameters"],
};

export const validateRequest = ajv.compile(deleteListRequestSchema);
