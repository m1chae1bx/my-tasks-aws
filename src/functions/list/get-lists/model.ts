import { JSONSchemaType, ajv } from "/opt/nodejs/util";

export interface GetListsRequest {
  pathParameters: {
    userId: string;
  };
}

export const getListsRequestSchema: JSONSchemaType<GetListsRequest> = {
  $schema: "http://json-schema.org/draft-07/schema#",
  type: "object",
  properties: {
    pathParameters: {
      type: "object",
      properties: {
        userId: {
          type: "string",
        },
      },
      required: ["userId"],
    },
  },
  required: ["pathParameters"],
};

export const validateRequest = ajv.compile(getListsRequestSchema);
