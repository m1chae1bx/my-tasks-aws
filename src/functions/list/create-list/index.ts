import { APIGatewayProxyEvent, APIGatewayProxyResult } from "aws-lambda";
import { validateRequest } from "./model";
import { List } from "/opt/nodejs/list.model";
import { genericErrorHandler } from "/opt/nodejs/util";

export const handler = async (
  event: APIGatewayProxyEvent
): Promise<APIGatewayProxyResult> => {
  try {
    const request = {
      pathParameters: event.pathParameters,
      body: event.body ? JSON.parse(event.body) : undefined,
    };

    if (!validateRequest(request)) {
      return {
        statusCode: 400,
        body: JSON.stringify({
          message: "Bad request",
          errors: validateRequest.errors,
        }),
      };
    }

    const userId = request.pathParameters.userId;
    const { name, isDefault } = request.body;

    const list = new List(name, userId, isDefault);
    const id = await list.save();
    const response = { message: `List ${name} was created successfully`, id };
    return {
      statusCode: 200,
      body: JSON.stringify(response),
    };
  } catch (error) {
    return genericErrorHandler(
      error,
      `An error occurred while creating the list. Please try again later.`
    );
  }
};
