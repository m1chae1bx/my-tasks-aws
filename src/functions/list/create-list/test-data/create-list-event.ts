import { baseAPIGatewayProxyEvent } from "@test-utils/base-objects";

export const createListEvent = {
  ...baseAPIGatewayProxyEvent,
  pathParameters: {
    userId: "testUserId",
  },
  body: JSON.stringify({
    name: "testListName",
    isDefault: false,
  }),
};
