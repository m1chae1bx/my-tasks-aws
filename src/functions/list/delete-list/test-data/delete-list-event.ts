import { baseAPIGatewayProxyEvent } from "@test-utils/base-objects";

export const deleteListEvent = {
  ...baseAPIGatewayProxyEvent,
  pathParameters: {
    userId: "testUserId",
    listId: "testListId",
  },
};
