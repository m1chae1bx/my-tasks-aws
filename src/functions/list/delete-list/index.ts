import { APIGatewayProxyEvent, APIGatewayProxyResult } from "aws-lambda";
import { validateRequest } from "./model";
import { List } from "/opt/nodejs/list.model";
import { genericErrorHandler } from "/opt/nodejs/util";

export const handler = async (
  event: APIGatewayProxyEvent
): Promise<APIGatewayProxyResult> => {
  try {
    const request = {
      pathParameters: event.pathParameters,
    };

    if (!validateRequest(request)) {
      return {
        statusCode: 400,
        body: JSON.stringify({
          message: "Bad request",
          errors: validateRequest.errors,
        }),
      };
    }
    const { userId, listId } = request.pathParameters;

    await List.delete(listId, userId);
    const response = {
      message: `List ${listId} was deleted successfully`,
    };
    return {
      statusCode: 200,
      body: JSON.stringify(response),
    };
  } catch (error) {
    return genericErrorHandler(
      error,
      `An error occurred while deleting the list. Please try again later.`
    );
  }
};
