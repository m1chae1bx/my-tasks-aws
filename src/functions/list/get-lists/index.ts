import { APIGatewayProxyEvent, APIGatewayProxyResult } from "aws-lambda";
import { validateRequest } from "./model";
import { List } from "/opt/nodejs/list.model";
import { genericErrorHandler } from "/opt/nodejs/util";

export const handler = async (
  event: APIGatewayProxyEvent
): Promise<APIGatewayProxyResult> => {
  try {
    const request = {
      pathParameters: event.pathParameters,
    };
    if (!validateRequest(request)) {
      return {
        statusCode: 400,
        body: JSON.stringify({
          message: "Bad Request",
          errors: validateRequest.errors,
        }),
      };
    }
    const userId = request.pathParameters.userId;

    const response = await List.getAll(userId);
    return {
      statusCode: 200,
      body: JSON.stringify(response),
    };
  } catch (error) {
    return genericErrorHandler(
      error,
      `An error occurred while getting the task lists. Please try again later.`
    );
  }
};
