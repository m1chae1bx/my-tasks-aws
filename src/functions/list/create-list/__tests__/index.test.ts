import { ajvError } from "@test-utils/base-objects";
import { APIGatewayProxyResult } from "aws-lambda";
import { handler } from "../index";
import { validateRequest } from "../model";
import { createListEvent } from "../test-data/create-list-event";
import { List } from "/opt/nodejs/list.model";
import { genericErrorHandler } from "/opt/nodejs/util";

jest.mock("../model", () => ({
  validateRequest: jest.fn(),
}));

jest.mock("/opt/nodejs/list.model", () => ({
  List: jest.fn(),
}));

jest.mock("/opt/nodejs/util", () => ({
  genericErrorHandler: jest.fn(),
}));

describe("create-list", () => {
  describe("happy path", () => {
    let result: APIGatewayProxyResult;

    beforeAll(async () => {
      (validateRequest as unknown as jest.Mock).mockReturnValueOnce(true);
      (List as unknown as jest.Mock).mockReturnValueOnce({
        save: jest.fn(),
      });
      result = await handler(createListEvent);
    });

    it("should return a 200 response", () => {
      expect(result.statusCode).toBe(200);
    });

    it("should return the list id", () => {
      expect(result.body).toMatchSnapshot();
    });
  });

  describe("sad path - invalid request", () => {
    let result: APIGatewayProxyResult;

    beforeAll(async () => {
      (validateRequest as unknown as jest.Mock).mockReturnValueOnce(false);
      validateRequest.errors = ajvError;
      result = await handler(createListEvent);
    });

    it("should return a 400 response", () => {
      expect(result.statusCode).toBe(400);
    });

    it("should return the validation errors", () => {
      expect(result.body).toMatchSnapshot();
    });
  });

  describe("sad path - invalid request (no body)", () => {
    let result: APIGatewayProxyResult;

    beforeAll(async () => {
      (validateRequest as unknown as jest.Mock).mockReturnValueOnce(false);
      validateRequest.errors = ajvError;
      result = await handler({ ...createListEvent, body: null });
    });

    it("should return a 400 response", () => {
      expect(result.statusCode).toBe(400);
    });

    it("should return the validation errors", () => {
      expect(result.body).toMatchSnapshot();
    });
  });

  describe("sad path - error while saving list", () => {
    let result: APIGatewayProxyResult;

    beforeAll(async () => {
      (validateRequest as unknown as jest.Mock).mockReturnValueOnce(true);
      (List as unknown as jest.Mock).mockReturnValueOnce({
        save: jest.fn().mockRejectedValueOnce(new Error("error")),
      });
      (genericErrorHandler as jest.Mock).mockReturnValueOnce({
        statusCode: 500,
        body: "test error",
      });
      result = await handler(createListEvent);
    });

    it("should return a 500 response", () => {
      expect(result.statusCode).toBe(500);
    });

    it("should return the error message", () => {
      expect(result.body).toMatchSnapshot();
    });
  });
});
