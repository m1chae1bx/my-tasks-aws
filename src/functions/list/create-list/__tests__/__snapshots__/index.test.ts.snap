// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`create-list happy path should return the list id 1`] = `"{\\"message\\":\\"List testListName was created successfully\\"}"`;

exports[`create-list sad path - error while saving list should return the error message 1`] = `"test error"`;

exports[`create-list sad path - invalid request should return the validation errors 1`] = `"{\\"message\\":\\"Bad request\\",\\"errors\\":[{\\"keyword\\":\\"\\",\\"instancePath\\":\\"\\",\\"schemaPath\\":\\"\\",\\"params\\":{},\\"message\\":\\"test error\\"}]}"`;
